projects/cds/src/merchandising/cms-components/merchandising-carousel/merchandising-carousel.component.ts

```diff
@@ -1,40 +1,73 @@
 import { ChangeDetectionStrategy, Component } from '@angular/core';
 import { Product } from '@spartacus/core';
 import { CmsComponentData } from '@spartacus/storefront';
-import { Observable, of } from 'rxjs';
+import { EMPTY, Observable, of } from 'rxjs';
 import {
   distinctUntilKeyChanged,
   filter,
   map,
   switchMap,
+  tap,
 } from 'rxjs/operators';
 import { CmsMerchandisingCarouselComponent } from '../../../cds-models/cms.model';
 import { CdsMerchandisingProductService } from '../../facade/cds-merchandising-product.service';
+import { MerchandisingProducts } from '../../model';
 
 @Component({
   selector: 'cx-merchandising-carousel',
   templateUrl: './merchandising-carousel.component.html',
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class MerchandisingCarouselComponent {
+  metadata: Map<string, string>;
+  items: Observable<Product>[];
+
   private componentData$: Observable<
     CmsMerchandisingCarouselComponent
   > = this.componentData.data$.pipe(filter(Boolean));
 
   title$: Observable<string> = this.componentData$.pipe(
     map(data => data.title)
   );
-  items$: Observable<Observable<Product>[]> = this.componentData$.pipe(
+
+  merchandisingProducts$: Observable<
```

Comment:
Instead of doing this workaround and creating side-effects in the stream, this might be a better alternative:

```ts

// note that it's private, we don't want to expose it
 private metadata$: Observable<Map<string, string>> = this.componentData$.pipe(
    map(data => {
      const metadata = new Map<string, string>()
      metadata.set('title', data.title);
      metadata.set('name', data.name);
      metadata.set('strategyid', data.strategy);
      metadata.set('id', data.uid);

      return metadata
    })
  )
  
  title$: Observable<string> = this.componentData$.pipe(
    map(data => data.title)
  );


  merchandisingProducts$: Observable<
    MerchandisingProducts
  > = this.componentData$.pipe(
    distinctUntilKeyChanged('strategy'),
    switchMap(data =>
      this.cdsMerchandisingProductService
        .loadProductsForStrategy(data.strategy, data.numberToDisplay)
    ),
    // this is no longer pipe-ing on the `loadProductsForStrategy` stream, it's a bit flatten.
    withLatestFrom(this.metadata$),
    map(([merchandsingProducts, metadata]) => {
      // TODO: set all the values from `metadata` variable to `merchandsingProducts.metadata`

      if (
        merchandsingProducts.products &&
        merchandsingProducts.products.length
      ) {
        merchandsingProducts.metadata.set(
          'slots',
          merchandsingProducts.products.length.toString()
        );
      }

      return merchandsingProducts;
    }),
    map(merchandsingProducts => {
      // TODO: introduce a new property `items` in `MerchandisingProducts` 
      // (or, alternatevly, create a new interface {merchandsingProducts: MerchandisingProducts, items: Observable<Product>[]})
      // and populate the `items` property using the logic from below.

     // NOTE: not sure if `Observable<Product>[]` is the appropriate type for `items`, please adjust if necessary.

      merchandisingProducts && merchandisingProducts.products
      ? merchandisingProducts.products.map(product => of(product))
      : [EMPTY];
    }),
    filter(Boolean)
  );
```

If you agree with above, and it works, you could try to maybe polish it more by moving the mapping logic to separate methods (similar like you had before, but now the methods are going to be called from the stream).

Reply:
As part of the future refactoring, we could consider moving this rather complex stream to a service, and make component cleaner and simpler.

Reply:
Updated to use this pattern.

I created the interface, but left it in this file as it isn't really an interface we want to expose widely. Happy to pull it out into its own file in the models if you would prefer.

Also the method to get the carousel metadata could have added the map to the `MerchandisingProducts` object rather than returning the map, but I felt it was more "functional" to not alter one of the provided parameters, again happy to change it if you think another way is better

Reply:
Good call 👍